File: Gems/Atom/Feature/Common/Assets/Scripts/material_property_overrides_demo.lua
----------------------------------------------------------------------------------------------------
--
-- Copyright (c) Contributors to the Open 3D Engine Project.
-- For complete copyright and license terms please see the LICENSE at the root of this distribution.
--
-- SPDX-License-Identifier: Apache-2.0 OR MIT
--
--
--
----------------------------------------------------------------------------------------------------

local PropertyValueTest = 
{
    Properties = 
    {
    }, 
}

function randomColor()
    return Color(math.random(), math.random(), math.random(), 1.0)
end

function randomDir()
    dir = {}
    for i = 1, 3 do
        lerpDir = math.random()
        if lerpDir < 0.5 then
            table.insert(dir, -1.0)
        else
            table.insert(dir, 1.0)
        end
    end
    return dir
end

function PropertyValueTest:OnActivate()
    self.timer = 0.0
    self.totalTime = 0.0
    self.totalTimeMax = 200.0
    self.timeUpdate = 2.0
    self.colors = {}
    self.lerpDirs = {}

    self.textures = 
    {
        "materials/presets/macbeth/05_blue_flower_srgb.tif.streamingimage",
        "materials/presets/macbeth/06_bluish_green_srgb.tif.streamingimage",
        "materials/presets/macbeth/09_moderate_red_srgb.tif.streamingimage",
        "materials/presets/macbeth/11_yellow_green_srgb.tif.streamingimage",
        "materials/presets/macbeth/12_orange_yellow_srgb.tif.streamingimage",
        "materials/presets/macbeth/17_magenta_srgb.tif.streamingimage"
    } 

    self.tickBusHandler = TickBus.Connect(self);
end

function PropertyValueTest:UpdateFactor(assignmentId)
    local propertyName = "baseColor.factor"
    local propertyValue = math.random()
    MaterialComponentRequestBus.Event.SetPropertyValue(self.entityId, assignmentId, propertyName, propertyValue);
end

function PropertyValueTest:UpdateColor(assignmentId, color)
    local propertyName = "baseColor.color"
    local propertyValue = color
    MaterialComponentRequestBus.Event.SetPropertyValue(self.entityId, assignmentId, propertyName, propertyValue);
end

function PropertyValueTest:UpdateTexture(assignmentId)
    if (#self.textures > 0) then
        local propertyName = "baseColor.textureMap"
        local textureName = self.textures[ math.random( #self.textures ) ]
        Debug.Log(textureName)
        local textureAssetId = AssetCatalogRequestBus.Broadcast.GetAssetIdByPath(textureName, Uuid(), false)
        MaterialComponentRequestBus.Event.SetPropertyValue(self.entityId, assignmentId, propertyName, textureAssetId);
    end
end

function PropertyValueTest:UpdateProperties()
    Debug.Log("Overriding properties...")
    for index = 1, #self.assignmentIds do
        local id = self.assignmentIds[index]
        if (id ~= nil) then
            self:UpdateFactor(id)
            self:UpdateTexture(id)
        end
    end
end

function PropertyValueTest:ClearProperties()
    Debug.Log("Clearing properties...")
    MaterialComponentRequestBus.Event.ClearAllPropertyValues(self.entityId);
end

function lerpColor(color, lerpDir, deltaTime)
    local lerpSpeed = 0.5
    color.r = color.r + deltaTime * lerpDir[1] * lerpSpeed
    if color.r > 1.0 then
        color.r = 1.0
        lerpDir[1] = -1.0
    elseif color.r < 0 then
        color.r = 0
        lerpDir[1] = 1.0
    end

    color.g = color.g + deltaTime * lerpDir[2] * lerpSpeed
    if color.g > 1.0 then
        color.g = 1.0
        lerpDir[2] = -1.0
    elseif color.g < 0 then
        color.g = 0
        lerpDir[2] = 1.0
    end

    color.b = color.b + deltaTime * lerpDir[3] * lerpSpeed
    if color.b > 1.0 then
        color.b = 1.0
        lerpDir[3] = -1.0
    elseif color.b < 0 then
        color.b = 0
        lerpDir[3] = 1.0
    end
end

function PropertyValueTest:lerpColors(deltaTime)
    for index = 1, #self.assignmentIds do
        local id = self.assignmentIds[index]
        if (id ~= nil) then
            lerpColor(self.colors[index], self.lerpDirs[index], deltaTime)
            self:UpdateColor(id, self.colors[index])
        end
    end
end

function PropertyValueTest:OnTick(deltaTime, timePoint)

    if(nil == self.assignmentIds) then
        local originalAssignments = MaterialComponentRequestBus.Event.GetDefaultMaterialMap(self.entityId)
        if(nil == originalAssignments or #originalAssignments <= 1) then -- There is always 1 entry for the default assignment; a loaded model will have at least 2 assignments
            return
        end
        
        self.assignmentIds = originalAssignments:GetKeys()

        for index = 1, #self.assignmentIds do
            local id = self.assignmentIds[index]
            if (id ~= nil) then
                self.colors[index] = randomColor()
                self.lerpDirs[index] = randomDir()
            end
        end
    end

    self.timer = self.timer + deltaTime
    self.totalTime = self.totalTime + deltaTime
    self:lerpColors(deltaTime)

    if (self.timer > self.timeUpdate and self.totalTime < self.totalTimeMax) then
        self.timer = self.timer - self.timeUpdate
        self:UpdateProperties()
    elseif self.totalTime > self.totalTimeMax then
        self:ClearProperties()
        self.tickBusHandler:Disconnect(self);
    end
end

return PropertyValueTest

File: Gems/Atom/Feature/Common/Assets/Scripts/material_property_overrides_demo.py
"""
Copyright (c) Contributors to the Open 3D Engine Project.
For complete copyright and license terms please see the LICENSE at the root of this distribution.

SPDX-License-Identifier: Apache-2.0 OR MIT
"""

import os
import sys
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
import azlmbr.atom
import azlmbr.math as math
import azlmbr.asset as asset
import azlmbr.bus as bus
import azlmbr.editor
import azlmbr.math as math
import random

from azlmbr.entity import EntityId
from azlmbr.render import MaterialAssignmentId
from azlmbr.name import Name

materials = [   'materials/presets/macbeth/00_illuminant.azmaterial',
                'materials/presets/macbeth/03_blue_sky.azmaterial',
                'materials/presets/macbeth/02_light_skin_tex.azmaterial',
                'materials/presets/macbeth/04_foliage_tex.azmaterial',
                'materials/presets/macbeth/07_orange.azmaterial',
                'materials/presets/macbeth/09_moderate_red_tex.azmaterial']

textures = [    'materials/presets/macbeth/05_blue_flower_srgb.tif.streamingimage',
                'materials/presets/macbeth/06_bluish_green_srgb.tif.streamingimage',
                'materials/presets/macbeth/09_moderate_red_srgb.tif.streamingimage',
                'materials/presets/macbeth/11_yellow_green_srgb.tif.streamingimage',
                'materials/presets/macbeth/12_orange_yellow_srgb.tif.streamingimage',
                'materials/presets/macbeth/17_magenta_srgb.tif.streamingimage']


def getRandomMaterial():
    material = random.choice(materials)
    return azlmbr.asset.AssetCatalogRequestBus(azlmbr.bus.Broadcast, 'GetAssetIdByPath', material, math.Uuid(), False)

def assetIdToPath(assetId):
    return azlmbr.asset.AssetCatalogRequestBus(azlmbr.bus.Broadcast, 'GetAssetPathById', assetId)

# List all slot ids
def getIds(entityId):
    print("Get material ids")
    return azlmbr.render.MaterialComponentRequestBus(azlmbr.bus.Event, 'GetDefaultMaterialMap', entityId).keys()

def printMaterials(entityId):
    materials = azlmbr.render.MaterialComponentRequestBus(azlmbr.bus.Event, 'GetMaterialMap', entityId)
    print(f"Materials: {[(id.ToString(), override.ToString()) for id, override in materials.items()]}", sep='\n')

def setMaterials(entityId, ids):
    print("Setting material overrides")
    for id in ids:
        azlmbr.render.MaterialComponentRequestBus(azlmbr.bus.Event, 'SetMaterialAssetId', entityId, id, getRandomMaterial())

# Sets overrides of 3 types: float, color, and texture
def setPropertyValues(entityId, ids):
    print("Setting property overrides")
    for id in ids:
        # factor override (float)
        factor = random.random()
        propertyName = "baseColor.factor"
        azlmbr.render.MaterialComponentRequestBus(azlmbr.bus.Event, 'SetPropertyValue', entityId, id, propertyName, factor)

        # color override
        color = math.Color(random.random(), random.random(), random.random(), 1.0)
        propertyName = "baseColor.color"
        azlmbr.render.MaterialComponentRequestBus(azlmbr.bus.Event, 'SetPropertyValue', entityId, id, propertyName, color)

        # texture override
        texturePath = random.choice(textures)
        assetId = azlmbr.asset.AssetCatalogRequestBus(azlmbr.bus.Broadcast, 'GetAssetIdByPath', texturePath, math.Uuid(), False)
        propertyName = "baseColor.textureMap"
        azlmbr.render.MaterialComponentRequestBus(azlmbr.bus.Event, 'SetPropertyValue', entityId, id, propertyName, assetId)

# Convert property value to string
def propertyValueToString(value):
    if isinstance(value, (int, float, bool, str)):
        return value
    if value.typename == 'Color':
        return 'Color(r={:.2f}, g={:.2f}, b={:.2f}, a={:.2f})'.format(value.r, value.g, value.b, value.a)
    if value.typename == 'Vector4':
        return 'Vector4(x={:.2f}, y={:.2f}, z={:.2f}, w={:.2f})'.format(value.x, value.y, value.z, value.w)
    if value.typename == 'Vector3':
        return 'Vector3(x={:.2f}, y={:.2f}, z={:.2f})'.format(value.x, value.y, value.z)
    if value.typename == 'Vector2':
        return 'Vector2(x={:.2f}, y={:.2f})'.format(value.x, value.y)
    if value.typename == 'ImageBinding':
        return assetIdToPath(value.GetAssetId())
    if value.typename == 'Asset':
        return assetIdToPath(value.GetAssetId())
    if value.typename == 'AssetId':
        return assetIdToPath(value)
    return value

# print all property overrides on a material component
def printPropertyValues(entityId, ids):
    print("Property overrides:")
    for id in ids:
        propertyOverrides = azlmbr.render.MaterialComponentRequestBus(azlmbr.bus.Event, 'GetPropertyValues', entityId, id)
        print(f"{id.ToString(), [(name.ToString(), propertyValueToString(override)) for name, override in propertyOverrides.items()]}", sep='\n')

# Clear property overrides for a single slot
def clearPropertyValues(entityId, id):
    print(f"Clearing property overrides for slot <{id.ToString()}>")
    azlmbr.render.MaterialComponentRequestBus(azlmbr.bus.Event, 'ClearPropertyValues', entityId, id)

# Clear all property overrides for the material
def clearAllPropertyValues(entityId):
    print("Clearing remaining property overrides")
    azlmbr.render.MaterialComponentRequestBus(azlmbr.bus.Event, 'ClearAllPropertyValues', entityId)

def run():
    searchFilter = azlmbr.entity.SearchFilter()
    searchFilter.names = ["Entity1"]
    matching_entities = azlmbr.entity.SearchBus(bus.Broadcast, 'SearchEntities', searchFilter)
    if matching_entities == 0:
        print('Could not find entity')
        return

    entityId = matching_entities[0]

    # Get ids
    ids = getIds(entityId)
    print(f"Slots: {[id.ToString() for id in ids]}", sep='\n')

    # Print current materials
    printMaterials(entityId)

    # Set material overrides
    setMaterials(entityId, ids)

    # Print current materials
    printMaterials(entityId)

    # Set property overrides
    setPropertyValues(entityId, ids)

    # Print current property overrides
    printPropertyValues(entityId, ids)

    # Clear property overrides for first slot
    for id in ids:
        clearPropertyValues(entityId, id)
        break

    # Print current property overrides
    printPropertyValues(entityId, ids)

    # Clear remaining property overrides
    clearAllPropertyValues(entityId)

    # Print current property overrides
    printPropertyValues(entityId, ids)

if __name__ == "__main__":
    run()
